Userland/Applications/PixelPaint/Tools/PolygonalSelectTool.cpp

```diff
@@ -0,0 +1,244 @@
+}
+
+void PolygonalSelectTool::on_mousemove(Layer*, MouseEvent& event)
+{
+    if (m_selecting)
+        m_last_selecting_cursor_position = event.layer_event().position();
+    m_editor->update();
+}
+void PolygonalSelectTool::on_doubleclick(Layer*, MouseEvent&)
+{
+    m_selecting = false;
+    m_editor->image().selection().end_interactive_selection();
+    process_polygon();
+    m_editor->did_complete_action(tool_name());
+    m_editor->update();
+}
+void PolygonalSelectTool::on_second_paint(Layer const* layer, GUI::PaintEvent& event)
+{
+    if (!m_selecting)
+        return;
+
+    GUI::Painter painter(*m_editor);
+    painter.add_clip_rect(event.rect());
+
+    painter.translate(editor_layer_location(*layer));
+
+    for (size_t i = 0; i < m_polygon_points.size() - 1; i++) {
+        auto preview_start = editor_stroke_position(m_polygon_points.at(i), 1);
+        auto preview_end = editor_stroke_position(m_polygon_points.at(i + 1), 1);
+        painter.draw_line(preview_start, preview_end, Color::Black, AK::max(m_editor->scale(), 1));
+    }
+
+    auto last_line_start = editor_stroke_position(m_polygon_points.at(m_polygon_points.size() - 1), 1);
+    auto last_line_stop = editor_stroke_position(m_last_selecting_cursor_position, 1);
+    painter.draw_line(last_line_start, last_line_stop, Color::Black, AK::max(m_editor->scale(), 1));
+}
+
+void PolygonalSelectTool::on_keydown(GUI::KeyEvent& key_event)
+{
+    Tool::on_keydown(key_event);
+    if (key_event.key() == KeyCode::Key_Escape) {
+        if (m_selecting) {
+            m_selecting = false;
+            m_polygon_points.clear();
+        } else {
+            m_editor->image().selection().clear();
+        }
+    }
+}
+
+GUI::Widget* PolygonalSelectTool::get_properties_widget()
+{
+    if (m_properties_widget)
+        return m_properties_widget.ptr();
+
+    m_properties_widget = GUI::Widget::construct();
+    m_properties_widget->set_layout<GUI::VerticalBoxLayout>();
+
+    auto& mode_container = m_properties_widget->add<GUI::Widget>();
+    mode_container.set_fixed_height(20);
+    mode_container.set_layout<GUI::HorizontalBoxLayout>();
+
+    auto& mode_label = mode_container.add<GUI::Label>();
+    mode_label.set_text("Mode:");
+    mode_label.set_text_alignment(Gfx::TextAlignment::CenterLeft);
+    mode_label.set_fixed_size(80, 20);
+
+    for (int i = 0; i < (int)Selection::MergeMode::__Count; i++) {
+        switch ((Selection::MergeMode)i) {
+        case Selection::MergeMode::Set:
+            m_merge_mode_names.append("Set");
+            break;
+        case Selection::MergeMode::Add:
+            m_merge_mode_names.append("Add");
+            break;
+        case Selection::MergeMode::Subtract:
+            m_merge_mode_names.append("Subtract");
+            break;
+        case Selection::MergeMode::Intersect:
+            m_merge_mode_names.append("Intersect");
+            break;
+        default:
+            VERIFY_NOT_REACHED();
+        }
+    }
```

Comment:
Note: You can make a lookup like this entirely at compile time like so. Which avoids allocating a vector and some strings.

```
static constexpr auto s_merge_mode_names = []{
    Array<StringView, (int)Selection::MergeMode::__Count> names;
    for (size_t i = 0; i < names.size(); i++) {
        switch((Selection::MergeMode)i) {
            case Selection::MergeMode::Set:
                names[i] = "Set";
                break;
             // ...
        }
    }
    return names;
}();
```

Where this is just placed somewhere in the .cpp.

Reply:
That's not true. `GUI::ItemListModel` is a template that can accept any container type, which by default is a `Vector`, and you've currently specialized it for `AK::String`. To initialize one using the static array of `StringView`s you'd do something like:

```c++
    auto model = GUI::ItemListModel<StringView, decltype(s_merge_mode_names)>::create(s_merge_mode_names);
```

(I'm using `decltype` since that's more convenient than writing the type)

See full example:  https://godbolt.org/z/Pb7YsnjrG

Reply:
Hey thanks! Can you tell I'm very very new to C++? :) I'll get this implemented this evening and push the changes.

Reply:
It's fine, this was more of an optional suggestion anyway :)

Reply:
This has been implemented, thanks again for the help getting it working.